graders/python/python_autograder/code_feedback.py

```diff
@@ -304,6 +304,110 @@ def bad(msg):
+                cls.add_feedback(msg)
+            if accuracy_critical:
+                cls.finish("")
+            else:
+                return False
+
+        if data is None:
+            return bad(f"{name} is None or not defined")
+
+        if not isinstance(data, dict):
+            return bad(f"{name} is not a dict")
+
+        # First, do all data type and partial keys checks.
+        if value_type is not None:
+            for value in data.values():
+                if not isinstance(value, value_type):
+                    return bad(
+                        f"{name} has the wrong type for value {value}, expected type {value_type}"
+                    )
+        if key_type is not None:
+            for key in data.keys():
+                if not isinstance(key, key_type):
+                    return bad(
+                        f"{name} has the wrong type for key {key}, expected type {key_type}"
+                    )
+
+        if partial_keys is None:
+            if check_values:
+                if len(ref.values()) != len(data.values()):
+                    return bad(
+                        f"{name} has the wrong number of values: expected {len(ref.values())}, got {len(data.values())}"
+                    )
+
+            if len(ref) != len(data):  # this is default length of keys check
+                return bad(
+                    f"{name} has the wrong number of entries: expected {len(ref)}, got {len(data)}"
+                )
+        check_partial_keys = partial_keys is not None and len(partial_keys) >= 1
+
+        # If any special checks enabled, do those
+        if check_keys or check_values or check_partial_keys:
+            # First, check for partial keys
+            if check_partial_keys:
+                for partial_key in partial_keys:
+                    if partial_key not in data:
+                        return bad(f"{name} does not contain key {partial_key}")
+
+            # Next, check that all keys are valid
+            if check_keys:
+                for key in data.keys():
+                    if key not in ref.keys():
+                        return bad(f"{name} contains an extra key: {key}")
+
+            # Finally, check all values are valid
+            if check_values:
+                if len(ref.values()) != len(data.values()):
+                    return f"{name} has the wrong length for values: expected {len(ref.values())}, got {len(data.values())}"
+                for value in data.values():
+                    if value not in ref.values():
+                        return bad(f"{name} contains an extra value: {value}")
+
+            # If all checks passed and we got to the end, return True
+            return True
+
+        # Otherwise, check equality of both keys and values between reference
+        # dict and student's dict
+        if ref == data:
+            return True
+
+        return bad(f"{name} has one or more key-value pairs that do not match")
```

Comment:
It might be nice to have options for more verbose error messages (key X did not match)

Reply:
I think that would be very useful in many situations 👍

Reply:
Fixed.